core/src/main/java/org/elasticsearch/cluster/routing/OperationRouting.java

```diff
@@ -54,16 +53,12 @@ public OperationRouting(Settings settings, AwarenessAllocationDecider awarenessA
         this.awarenessAllocationDecider = awarenessAllocationDecider;
     }
 
-    public ShardIterator indexShards(ClusterState clusterState, String index, String type, String id, @Nullable String routing) {
-        return shards(clusterState, index, type, id, routing).shardsIt();
+    public ShardIterator indexShards(ClusterState clusterState, String index, String id, @Nullable String routing) {
```

Comment:
I think we can reduce the scope of this change by exposing a resolveShardId method that resolves index,type,id and routing to a shardId. And then we don't need to touch these. Also, I know that type is not used now, but why not pass it? is there a place we don't have it? I hope we can back port this change to 2.x, so having type here will reduce the change.